### PR TITLE
[Profiler] Map regions for error-throwing AST nodes

### LINKED ISSUE #61318: Incorrect coverage counters following a `catch` block with a `return`
For the following:

```swift
enum E: Error { case e }

func foo() throws { throw E.e }
func bar() -> Int {
  do {
    try foo()
    return 1
  } catch {
    return 2
  }
}

_ = bar()
```

the line coverage comes out as:

```
    1|func foo() throws { throw E.e }
    1|func bar() -> Int {
    1|  do {
    1|    try foo()
    1|    return 1
    1|  } catch {
    1|    return 2
18.4E|  }
18.4E|}
     |
    1|_ = bar()
```

we're somehow ending up with counters subtracting from zero and underflowing.

rdar://100470244

### LINKED ISSUE #63194: Code coverage reports a line as being covered when it isn't.
**Description**
With code coverage, lines that aren’t executed after a throw still have their hit count incremented.

**Steps to reproduce**
1. Open the sample project
2. Run testExample() unit test
3. Notice in Throws.swift that the commented line shows 3 hits instead of 1

**Expected behavior**
The commented line should show 1 hit

**Environment**
- Swift compiler version info: swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
Target: arm64-apple-macosx13.0

- Xcode version info: Version 14.2 (14C18)
- Deployment target: macOS 13.1

Also reported as: FB7656633

[CoverageBug.zip](https://github.com/apple/swift/files/10492831/CoverageBug.zip)

### PR DESCRIPTION
**Pull request by hamishknight**
Map a counter for the error branch of a given potentially-throwing expression, and subtract it from the following region count. This requires also fixing the coverage for do-catch statements, which would have otherwise been regressed.

rdar://34244637
rdar://100470244
rdar://118185163
Resolves #63194
Resolves #61318

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please build toolchain macOS

File: test/Profiler/coverage_errors.swift [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+  do {                      // CHECK-NEXT: [[@LINE]]:6  -> [[@LINE+2]]:4 : 0
+    return try throwingFn() // Note we don't emit a region here because it would be empty.
+  } catch {                 // CHECK-NEXT: [[@LINE]]:11 -> [[@LINE+2]]:4 : 2
+    return 0
+  }
+}                           // CHECK-NEXT: }
+
+func takesInts(_ x: Int, _ y: Int) {}
+
+// The throwing expr is nested here, the region starts after the throwing expr.
```
Comment (bnbarham) [inline test/Profiler/coverage_errors.swift:314]: Another to consider here would be eg.
```
func foo() throws -> Int { return  1}
func foo2(_ arg: Int, _ arg2: Int, _ arg3: Int) throws -> Int { return 2 }
func bar() throws -> Int {
  // code here
  try foo2(
    1,
    foo(),
    2
  )
  // more here
}
```

The call to `foo2` isn't executed if `foo()` throws. Whether that's worth differentiating... 🤷

EDIT: Ah, you have this case further down in `test20`, could be worth a TODO there.

**Review by bnbarham — APPROVED** [review]

File: lib/SIL/IR/SILProfiler.cpp [hunk 2]
```diff
@@ -173,7 +163,15 @@ SILLocation ProfileCounterRef::getLocation() const {
...
-  case Kind::Node: {
+  case Kind::Node:
+    break;
+  case Kind::ErrorBranch:
+    OS << "error branch of: ";
+    break;
+  }
+  switch (RefKind) {
+  case Kind::Node:
+  case Kind::ErrorBranch: {
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/SIL/IR/SILProfiler.cpp:174]: Yeah

File: lib/SIL/IR/SILProfiler.cpp [hunk 3]
```diff
@@ -342,168 +367,70 @@ struct MapRegionCounters : public ASTWalker {
     if (isa<LazyInitializerExpr>(E))
       mapRegion(E);
 
-    return shouldWalkIntoExpr(E, Parent, Constant);
+    auto WalkResult = shouldWalkIntoExpr(E, Parent, Constant);
+    if (WalkResult.Action.Action == PreWalkAction::SkipChildren) {
+      // We need to manually do the post-visit here since the ASTWalker will
+      // skip it.
+      // FIXME: The ASTWalker should do a post-visit.
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/SIL/IR/SILProfiler.cpp:374]: Yeah :/

File: test/Profiler/coverage_errors.swift (hunk 1 above)
Comment (hamishknight) [reply test/Profiler/coverage_errors.swift:314]: Yeah I'll look into this as a follow up

**Pull request merged**